compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp

```diff
@@ -104,6 +104,77 @@ static FailureOr<SmallVector<Value>> rewriteAsPaddedOp(
   return paddedSubviewResults;
 }
 
+static FailureOr<Value> rewriteAsPaddedOp(IRRewriter &rewriter,
+                                          tensor::UnPackOp op,
+                                          tensor::UnPackOp &paddedOp) {
+  Location loc = op.getLoc();
+
+  // Set IP after op because we also take the dims of the original output.
+  IRRewriter::InsertionGuard g(rewriter);
+  rewriter.setInsertionPointAfter(op);
+
+  auto tensorLoad =
+      op.getDest().getDefiningOp<IREE::Flow::DispatchTensorLoadOp>();
+  if (!tensorLoad) {
+    return rewriter.notifyMatchFailure(op, "does not have tensor load");
+  }
+
+  // Determine the padded shape from the load
+  ArrayRef<int64_t> shape = op.getDestType().getShape();
+  SmallVector<int64_t> paddedShape(shape.begin(), shape.end());
+  for (const auto &[index, size] :
+       llvm::enumerate(tensorLoad.getMixedSizes())) {
+    if (Optional<int64_t> cst = getConstantIntValue(size)) {
+      paddedShape[index] = cst.value();
+    } else {
+      FailureOr<int64_t> upperBound =
+          linalg::getConstantUpperBoundForIndex(size.get<Value>());
+      if (failed(upperBound)) {
+        return rewriter.notifyMatchFailure(
+            op, "No constant bounding box can be found for padding");
+      }
+      paddedShape[index] = *upperBound;
+    }
+  }
```

Comment:
can we refactor that a bit? Looks like exact same code as the function above?

Reply:
SG, done!